src/docker/idle_container_creator.rs

```diff
@@ -0,0 +1,138 @@
+use std::sync::mpsc::{sync_channel, Receiver};
+use std::thread;
+use std::time::Duration;
+
+use lazy_static::lazy_static;
+use parking_lot::Mutex;
+use subprocess::Popen;
+use tempfile::TempDir;
+
+use crate::docker::call_docker_async;
+use crate::error::{WorkerError, WorkerErrorKind};
+use crate::fs_utils::get_filename;
+use crate::named_pipe::NamedPipe;
+
+// NOTE: the number of idle containers on the system is CONTAINER_CACHE_CHANNEL_SIZE + CACHE_POPULATOR_COUNT
+const CONTAINER_CACHE_CHANNEL_SIZE: usize = 3;
+const CACHE_POPULATOR_COUNT: usize = 2;
+
+const CONTAINER_IMAGE_TAG: &str = "python";
+// 1000000000 seconds ~= 30 years
+const SLEEP_TIME: &str = "1000000000";
+
+const CONTAINER_PIPE_FOLDER: &str = "/home/io";
+
+fn next_container_name() -> String {
+    let id: u64 = rand::random();
+    format!("v9_{}_{}", CONTAINER_IMAGE_TAG, id)
+}
+
+fn sync_create_container() -> Result<Container, WorkerError> {
+    let pipe_dir = TempDir::new()?;
+    let pipe_path = pipe_dir.path().canonicalize()?;
+    let pipe_path_string = pipe_path
+        .into_os_string()
+        .into_string()
+        .map_err(WorkerErrorKind::OsStringConversion)?;
+    let pipe_mount = format!("{}:{}", pipe_path_string, CONTAINER_PIPE_FOLDER);
+
+    let desired_name = next_container_name();
+    let docker_subprocess = call_docker_async(&[
+        "run",
+        "-v",
+        &pipe_mount,
+        "--name",
+        &desired_name,
+        CONTAINER_IMAGE_TAG,
+        "sleep",
+        SLEEP_TIME,
+    ])?;
+
+    let pipe = NamedPipe::in_dir(pipe_dir)?;
+
+    Ok(Container {
+        pipe,
+        docker_container_name: desired_name,
+        docker_run_process: docker_subprocess,
+    })
+}
+
+pub struct IdleContainerCreator {
+    cache_channel_receiver: Mutex<Receiver<Container>>,
+}
+
+impl IdleContainerCreator {
+    fn new() -> Self {
+        // Create the cache channel
+        let (sender, receiver) = sync_channel(CONTAINER_CACHE_CHANNEL_SIZE);
```

Comment:
I should probably look this up, but what is a cache channel?

Reply:
https://doc.rust-lang.org/std/sync/mpsc/fn.sync_channel.html

We just use it to store cached containers